static/app/components/devtoolbar/hooks/useReplayRecorder.tsx

```diff
@@ -0,0 +1,116 @@
+}
+
+function getReplayInternal(
+  replay: ReturnType<typeof replayIntegration>
+): ReplayInternalAPI {
+  // While the toolbar is internal, we can use the private API for added functionality and reduced dependence on SDK release versions
+  // @ts-ignore:next-line
+  return replay._replay;
+}
+
+const LAST_REPLAY_STORAGE_KEY = 'devtoolbar.last_replay_id';
+
+export default function useReplayRecorder(): ReplayRecorderState {
+  const {SentrySDK} = useConfiguration();
+  const replay =
+    SentrySDK && 'getReplay' in SentrySDK ? SentrySDK.getReplay() : undefined;
+  const replayInternal = replay ? getReplayInternal(replay) : undefined;
+
+  // sessionId is defined if we are recording in session OR buffer mode.
+  const [sessionId, setSessionId] = useState<string | undefined>(() =>
+    replayInternal?.getSessionId()
+  );
+  const [recordingMode, setRecordingMode] = useState<ReplayRecordingMode | undefined>(
+    () => replayInternal?.recordingMode
+  );
+
+  const isDisabled = replay === undefined;
+  const disabledReason = !SentrySDK
+    ? 'Failed to load the Sentry SDK.'
+    : !('getReplay' in SentrySDK)
+      ? 'Your SDK version is too outdated to use the Replay integration.'
+      : !replay
+        ? "Failed to load your SDK's Replay integration."
+        : undefined;
+
+  const [isRecording, setIsRecording] = useState<boolean>(
+    () => replayInternal?.isEnabled() ?? false
+  );
+  const [lastReplayId, setLastReplayId] = useState<string | undefined>(
+    () => sessionStorage.getItem(LAST_REPLAY_STORAGE_KEY) || undefined
+  );
+  useEffect(() => {
+    if (isRecording && sessionId) {
+      setLastReplayId(sessionId);
+      sessionStorage.setItem(LAST_REPLAY_STORAGE_KEY, sessionId);
+    }
+  }, [isRecording, sessionId]);
+
+  const refreshState = useCallback(() => {
+    setIsRecording(replayInternal?.isEnabled() ?? false);
+    setSessionId(replayInternal?.getSessionId());
+    setRecordingMode(replayInternal?.recordingMode);
+  }, [replayInternal]);
+
+  const start = useCallback(async () => {
+    let success = false;
+    if (replay && !isRecording) {
+      try {
+        // SDK v8.19.0 and older will throw if a replay is already started.
+        // Details at https://github.com/getsentry/sentry-javascript/pull/13000
+        if (recordingMode === 'session') {
+          replay.start();
+        } else {
+          // For SDK v8.20.0 and up, flush() works for both cases.
+          await replay.flush();
+        }
+        success = true;
+        // eslint-disable-next-line no-empty
+      } catch {}
+    }
+    refreshState();
+    return success;
+  }, [isRecording, recordingMode, replay, refreshState]);
+
+  const stop = useCallback(async () => {
+    let success = false;
+    if (replay && isRecording) {
+      try {
+        await replay.stop();
+        success = true;
+        // eslint-disable-next-line no-empty
+      } catch {}
+    }
+    refreshState();
+    return success;
```

Comment:
Could be a good candidate for a [`finally` block](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#the_finally_block)

Reply:
I'd like it to cover the outer condition too

Reply:
```suggestion
try {
  if (replay && isRecording) {
    await replay.stop();
    return true;
  }
  return false;
} catch {
  return false;
} finally {
  refreshState();
}
```